### DISCUSSION TITLE
Gentoo build

### DISCUSSION THREAD

**Discussion by immolo** [frno7/mipsr5900el-gentoo-linux-gnu#17] [General]
Well it's been a long weekend of debugging but at least some progress is happening:

![running](https://user-images.githubusercontent.com/78145083/169716054-dd424a6f-0cf5-43d8-a908-0aad5714793a.png)

Just a`switch_root` kernel panic issue to go now and we'll have a shiny Gentoo USB to play with :)

I'll drop an update when I return from work next week.

**Comment by frno7** [comment]
What’s up with the `switch_root` kernel panic?

**Reply by immolo** [reply]
We haven't managed to find what is causing it to be able to answer this as all the debugging shows it should be working.

Luckily the Gentoo Community is amazing so we have many eyes looking into this, so someone will catch the issue.

**Reply by immolo** [reply]
`
#!/bin/busybox sh

/bin/busybox --install -s

PATH=/usr/sbin:/usr/bin:/sbin:/bin
export PATH

#Mount Section

mount -t proc     none /proc
mount -t devtmpfs none /dev
mount -t sysfs    none /sys


#Load PS2 Modules

depmod

modprobe ps2fb mode_option=640x512i@50
modprobe sif
modprobe iop-memory
modprobe iop-module
modprobe iop-irq

modprobe sd_mod
modprobe ohci-ps2
modprobe ums-usbat
modprobe usbhid
modprobe hid-generic
sleep 5

#Mount USB
mount -t ext4 -o ro /dev/sda2 /mnt || echo "Failed to mount sda2"
sleep 30

#Debug

#/bin/sh
mkdir -p /mnt/proc
mount -o move /proc /mnt/proc
mkdir -p /mnt/sys
mount -o move /sys /mnt/sys
mkdir -p /mnt/dev
mount -o move /dev /mnt/dev

echo "this is '$$'"

exec switch_root /mnt /sbin/init`

This is the init we are currently using.

If you want to try yourself then this setup expects a USB with a fat32 partition for the kernel at sda1 and the unpacked stage3 in ext4 at sda2. Screenshots of your findings will help me if you can share.

**Reply by frno7** [reply]
> Also I've discovered that once the initramfs is discarded it doesn't change the allocation of RAM so the 5.4 kernel definitely has 9mb lower than the 2.6 kernel.

[Kexec](https://en.wikipedia.org/wiki/Kexec) may be a convenient mechanism to reduce kernel memory. :-)

**Reply by immolo** [reply]
I didn't even know this could be used like this so thanks for the hint!

I'll be testing the 23.0 profiles soon on mips and that's when nearly all known musl gentoo mips issues should hopefully be solved so I can use that as an excuse to spend more time on the PS2 looking for bugs so I'll post some updates later on in the month on it until the profiles are stable and I can release something stable.

Also happy new years too you! Thanks for all the little lessons this year as your help and explanations have improved my skills enough that I've been able to spot and submit fix issues to make Gentoo on MIPS better for everyone, which isn't isn't a bad outcome from just wanting to mess with Gentoo on my PS2 because I was bored one day :)

**Reply by frno7** [reply]
Yeah, happy new year, and thanks for all Gentoo fixes!

**Comment by frno7** [comment]
> If you want to try yourself then this setup expects a USB with a fat32 partition for the kernel at sda1 and the unpacked stage3 in ext4 at sda2.

Oh, does it need two USB drive partitions? An alternative could be to have a single [FAT32](https://en.wikipedia.org/wiki/FAT32) partition with two files, one for the kernel and another for [SquashFS](https://en.wikipedia.org/wiki/SquashFS), containing the stage3, and mount it as a [loop device](https://en.wikipedia.org/wiki/Loop_device). For most people, that’d be much easier to install, especially with non-Linux operating systems. :-)

**Reply by immolo** [reply]
I thought about this but then remembered not everyone has a hard drive so my solution means you can either continue installing on this USB image or install directly to hard drive if you have one. I'm trying to think of this from an end users viewpoint and only having 2 USB ports.

dd or win32imager means this is no drama for Linux or other OS users.

**Reply by frno7** [reply]
Both variants could, potentially, be detected and handled accordingly by the same `/sbin/init` script. I’m thinking of the casual person just wanting to boot Gentoo on the PlayStation 2 to see what it looks like, which is why I thought copying two files to a preexisting FAT partition would be simplest. Anyhow, let’s proceed with your suggestion. :-)

**Reply by immolo** [reply]
Sorry for the slow update but life gets in the way of the fun sometimes.

I've got Gentoo to boot now so I'm going to use my long weekend to optimise the build and clean up the initramfs so it works automagically rather than just for my system. Once this this is done I'll turn my notes into a wiki install guide which I'd be grateful if you all could follow to make sure we have caught every issue a new user will have between us :)

@frno7 Have you done any tests on cflags yet as I trust you enough with Gentoo that I can skip a few hours of work running benchmarks to find them myself.

**Reply by frno7** [reply]
Hmm, I don’t think I’ve had any special `CFLAGS` for the R5900, except for the `-mfix-r5900` option, but with an `mipsr5900el-unknown-linux-` target it’s automatic. It’s only needed for non-R5900 compiler targets, due to the [R5900 short loop erratum](https://github.com/frno7/linux/wiki/R5900-short-loop-erratum). Issue https://github.com/frno7/linux/issues/3 will be a major optimisation later on, and some of the other issues. In general: functionality, robustness, and usability seem more important than speed at this stage. I think we can relax and take it easy with `CFLAGS`, unless someone likes to fiddle with it, of course. :-)

**Reply by immolo** [reply]
Looks like some exciting stuff coming up in the kernel for me to play with then :)

I'm one of those people that gets their kicks from pushing systems to run as far as possible before it crashes so even though my head agrees with you, the masochist in me knows how it will end.

I've also started putting the feelers out for us to get access to a server that can build and host a binary host for us so we may actually have a chance to set this up the right way from day one.

**Reply by immolo** [reply]
Ok so I've dropped an alpha image for everyone to have a play with as this is fun and wrote a small guide in the wiki - https://github.com/frno7/gentoo-mipsr5900el/wiki/Gentoo-for-the-PlayStation-2

I have some work to do with getting networking started but first I need to get Gentoo running from the hard drive to make debugging easier. Is the pata_platform module all you need to get hard drives running from the expansion bay?

**Reply by AKuHAK** [reply]
> pata_platform

ATA registers is not implemented, so cirrently only usb devices are working.

**Reply by AKuHAK** [reply]
https://github.com/frno7/linux/issues/18

**Reply by immolo** [reply]
@AKuHAK Oh I read that issue which got me on to the pata_platform in the first place with frno7 showing mounting of drive and benchmarking.

**Reply by frno7** [reply]
The ATA driver is in preparation. It does reading via DMA (mostly) fine, but needs refactoring to support writing via DMA in a sensible manner. The games controller driver is also being prepared, https://github.com/frno7/linux/issues/22. As a prototype, it’s already quite usable with the DualShock 2. I’ve also made a bit of progress on supporting a [virtual frame buffer](https://github.com/frno7/linux/wiki/The-Graphics-Synthesizer#virtual-frame-buffer).

**Reply by immolo** [reply]
That's a shame then however I'll solider on with those 11mbps transfer speeds.

Also I've discovered that once the initramfs is discarded it doesn't change the allocation of RAM so the 5.4 kernel definitely has 9mb lower than the 2.6 kernel.

**Reply by frno7** [reply]
> Also I've discovered that once the initramfs is discarded it doesn't change the allocation of RAM so the 5.4 kernel definitely has 9mb lower than the 2.6 kernel.

Oh, not _definitely_. There are at least three options remaining to reclaim most if not all of that RAM:

1. figure out if initramfs can be reclaimed by some reasonable method;
2. make a custom initramfs having only the logic needed to support the live USB. I suppose this initramfs could be made as small as 4 KiB, or so, which would be more than 1000 times smaller than now;
3. hard-code reasonable kernel parameters and remove initramfs altogheter.

Options (1) and (2) can be combined to great effect, of course. The main reason to have an initramfs is that it can do scripting, for example dynamically determine the appropriate partition to boot, and so on.

**Reply by immolo** [reply]
OK poor choice of word on my part :)

I'm not quite sure on dropping the initramfs to 4k but I could save a few mb off from ideas you have just given me. I think my play though is to hardcode a kernel and confirm the situation is happening like I think it is then as you say try and figure an option 1 and 2 final solution.

**Reply by frno7** [reply]
Removing all _unused_ kernel modules from `/lib/modules/5.4.169+/kernel` should save more than 4 MB of RAM in one fell swoop. :-)

**Reply by immolo** [reply]
@frno7 That was literally what came to mind when I read your comment :)

In other news networking and ssh is now fully working on my machine so it's time to start looking into MUSL now I have something to compare then I can start looking into freeing up memory.

**Reply by frno7** [reply]
Nice! Is networking Ethernet by USB?

**Reply by immolo** [reply]
Yes, I have an old USB ethernet adapter which wouldn't work yesterday but magically is today so I need to wait for some user reports to come to see if I need to look further into this.

**Reply by frno7** [reply]
> Yes, I have an old USB ethernet adapter which wouldn't work yesterday but magically is today so I need to wait for some user reports to come to see if I need to look further into this.

Insufficient USB buffer memory is a likely cause of troubles, including erratic (or magical) behaviour and apparent [deadlocks](https://en.wikipedia.org/wiki/Deadlock) as devices contend over resources. Memory pressure increases with the number of attached devices. Wifi drivers tend to demand a lot, for instance. Sometimes their requirements can be dialled down to work, as in commit https://github.com/frno7/linux/commit/844ad983a20ea374589ffc033933c0dedd65bc70. There’s a brief [USB article](https://github.com/frno7/linux/wiki/USB-for-the-PlayStation-2) about this on the wiki. Modern computers supply magnitudes more USB buffer memory than the current 256 KiB hard limit total of the PlayStation 2 kernel, so the likelihood of triggering stuff never seen before elsewhere is much greater.

**Reply by immolo** [reply]
I didn't even think of this so those issues have helped me picture this in my head.

My current theory is something strange is happening when the device is first time connecting to a router for a dhcp IP but if you have it assigned on to it already then it works. I don't have enough information on this at the moment though so I'm waiting for feedback from people testing this on Reddit so I have enough information.

I have a selection of WiFi chipsets in my drawer though so that will be fun to go through and see which ones can be made to work now it's easier to have an environment running.

**Reply by AKuHAK** [reply]
@immolo can you post the link to reddit?

**Reply by immolo** [reply]
> @immolo can you post the link to reddit?

@AKuHAK sure, https://www.reddit.com/r/linux/comments/v6dzjl/well_if_we_are_having_a_trend_for_linux_on_crazy/

All issues should be coming back to here rather than there though so we can all use the information these people provide us.

**Reply by frno7** [reply]
> I didn't even think of this so those issues have helped me picture this in my head.

Thanks for your USB observation. Hopefully these things will be explained in wiki articles!

> My current theory is something strange is happening when the device is first time connecting to a router for a dhcp IP but if you have it assigned on to it already then it works. I don't have enough information on this at the moment though so I'm waiting for feedback from people testing this on Reddit so I have enough information.

Oh.

> I have a selection of WiFi chipsets in my drawer though so that will be fun to go through and see which ones can be made to work now it's easier to have an environment running.

Go for it. :-) There are at least two options to improve the USB situation:

1. Increase the [USB buffer size](https://github.com/frno7/linux/blob/59a11ab94a4020408c7dfbf92bd55778e17b43f1/drivers/usb/host/ohci-ps2.c#L24-L25), doubling it to 512 KiB for instance, with this define:

```C
/* Size allocated from IOP heap (maximum size of DMA memory). */
#define DMA_BUFFER_SIZE (256 * 1024)
```

However, bear in mind that the IOP typically has 2 MiB in total, and it’s supposed to (eventually) perform other services, such as ATA, Ethernet, audio, etc. that need buffers too. The technique is called _bounce buffer_, and it’s seldom used with the Linux kernel USB subsystem. Rare bugs can and sometimes do creep up with this combination, as I’ve fixed a few already. Its raison d’être being that the USB DMA hardware is only available with (2 MiB) IOP memory, as opposed to (32 MiB) main memory.

2. Design a more advanced bounce buffer, or make some other kind of USB driver adjustment.

**Reply by immolo** [reply]
@frno7 just the explanation I needed!

I'm getting to the stage where I have so much I want to try but not enough time in the day.

I'm currently getting the framework in place for the binhost as I did an experiment on how bad compiling would be and dhcpcd took 10 hours to build so I'll never do that again...

**Reply by frno7** [reply]
> I'm currently getting the framework in place for the binhost as I did an experiment on how bad compiling would be and dhcpcd took 10 hours to build so I'll never do that again...

Compilation time is obviously an issue. For complex packages GCC demands gigabytes of memory that, with [swapping](https://en.wikipedia.org/wiki/Memory_paging), probably would take years or decades to compile on PlayStation 2 hardware. Compiling GCC with GCC itself, for instance.

**Reply by immolo** [reply]
I build Gentoo on weak hardware for fun so I needed a baseline of just how bad it was going to be for how many packages are going to be needed. After this though I'm starting to think we are going to need to carry a lot more than I originally thought.

As a side effect though we now have a method to benchmark optimisation you make in the kernel ;)

**Reply by frno7** [reply]
I’m handwaving a bit here, as a starting point for thought: given 20+ years of processor hardware design advances since the 300 MHz R5900 was introduced, maybe a modern core is 5 times faster clock-for-clock. Multiply that with 10 for a faster (3 GHz) clock, and 10 cores of those, and we have a 500 times faster modern computer, roughly. On top of that there’s mass storage performance to consider, and, of course, if it runs out of memory and starts paging, it’ll be 1E3 to 1E6 times slower still. Worst case, that’s roughly 1E9, or a billion, times slower than a modern processor with a sufficient amount of memory.

**Reply by immolo** [reply]
@frno7 I read that and just thought how much I want to go for a beer and listen to your mind explain the things I love in this level of detail, please never change :)

I'm probably not making myself clear though as I often skips steps when I'm excited about a new challenge to solve, so in my own simpler understanding I have a Gentoo system on a 600mhz Pentium 3 which I use as a baseline for old hardware projects and with the USB1.1 USB root dhcpcd was taking 30 minutes to build and GCC 11 was taking 7 days. So when I was saying that I was surprised with the 10 hours of dhcpcd it was because I had 2 hours as a reasonable guess with the knowledge I had, plus I've done a stage1 install on the PS2 in the past so even though I know code is more complex now it always surprises me by just how much it is.

I forgot to ask earlier as well to which I'm sorry for doing it without checking with you first, but are you happy if Gentoo PS2 bugs end up being posted in this issue repo as not only does it keep all the work together for a more community feel but it also means I get to keep the fantastic advice from yourself and @AKuHAK that so far has always managed to kick start me into the right direction. I also think it will find kernel issues quicker too as you both have a deeper understanding of the system to identify them compared to me.

As my goal was to take the system side off your hands so you could spend more time on the kernel side, then in exchange for this I will also take the role of closing and moving those bugs upstream when needed if you wish so I'm not putting more work on your lap. Unless you both have a better idea that is?

**Reply by frno7** [reply]
Let’s try your ideas!

> So when I was saying that I was surprised with the 10 hours of dhcpcd it was because I had 2 hours as a reasonable guess with the knowledge I had, plus I've done a stage1 install on the PS2 in the past so even though I know code is more complex now it always surprises me by just how much it is.

Performance deteriorates _very_ quickly into an almost complete stand-still if the [working set](https://en.wikipedia.org/wiki/Working_set) (in this case mainly GCC doing its things) is greater than available RAM. For example, if, say, 90 % of RAM is actively used for a given working set at one particular time, performance might be reasonable, and somewhat easy to anticipate. If, for whatever reason (two rather than one active SSH connection, or what not happening in the background), a working set tips over to 110 % of available RAM, or worse, that often leads to dramatic effects and, essentially, arbitrarily low performance, because Linux is forced to swap parts of the working set via a mass storage device. So, one compilation run might be 10 times faster only because it had a MB or so of extra RAM available for its working set than some other run.

Compiling for 10 hours straight on PlayStation 2 hardware with a recent 5.x kernel must surely be some kind of record, and a great test of overall system reliability. Thanks! :-)

**Comment by immolo** [comment]
@frno7 Sorry to bother you with something mundane however I'm in the process of having the MUSL patch accepted into the main Gentoo tree until it gets accepted upstream and need I to double check if it's failing over an issue with the patch being applied to MUSL 1.2.3 or if I just can't see the issue with my ebuild for it.

[ebuild](https://dpaste.com/2L5ATW5TF)
[log](https://dpaste.com/ER9A76WAG)

There is also a small chance my host system is playing up from abusing my Arch install over the last 4 years so I just need the eyes of someone that understands the whole process as currently it should be working but it's clearly not.

**Reply by frno7** [reply]
Your links don’t seem to work, but before debugging your ebuild you might want to try the three commands related to Musl as used to produce the Docker for this repo:

https://github.com/frno7/gentoo-mipsr5900el/blob/beecdfd85a13c687a5b8428d61076b954d87362d/scripts/install-crossdev#L13-L17

**Reply by immolo** [reply]
Must have expired however I've got it build manually so it's just an issue with eapply I need to figure out.

**Reply by immolo** [reply]
Turns out it was my host system crying from the abusing I put it under so give me a few days for an update and hopefully bug report to track the progress of your patch being added.

How did you submit the patch to the main MUSL team as well? I may as well try and push that on too while I'm in a bug hunting mood :)

**Reply by frno7** [reply]
Oh, see link to Musl patch post in https://github.com/frno7/linux/issues/33#issuecomment-750902872.

**Reply by immolo** [reply]
Ok, my cross compiler for MUSL is built although I did run into a strange issue with not being able to build 32bit code in my 64bit chroot which was really odd and something future me needs to take a look at later on.

https://bugs.gentoo.org/851759

**Reply by frno7** [reply]
Lovely. :-) To my understanding Musl/MIPS is o32 only, but I haven’t seen any issues building it with crossdev on a 64-bit host.

**Reply by immolo** [reply]
I'm speaking to the lead devs about it as I think I may have run in to a bug for upstream.

Are you using MUSL 1.2.3 and GCC 11.3 in the docker and your host? because if you aren't then I can just submit the bug right now as we have ruled everything else out.

**Reply by immolo** [reply]
@frno7 if you could share the output of your `emerge --info`as well that will help pinpoint if we are doing something different.

**Reply by frno7** [reply]
I’m not sure where the Docker is sitting at the moment, but I updated my workstation yesterday and it’s got Musl 1.2.3 with GCC 11.3.0 for sure:

```
% mipsr5900el-unknown-linux-musl-gcc --version
mipsr5900el-unknown-linux-musl-gcc (Gentoo 11.3.0 p4) 11.3.0
Copyright (C) 2021 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
% /usr/mipsr5900el-unknown-linux-musl/usr/lib/libc.so
musl libc (mipsel-sf)
Version 1.2.3
Dynamic Program Loader
Usage: /usr/mipsr5900el-unknown-linux-musl/usr/lib/libc.so [options] [--] pathname [args]
```

**Reply by immolo** [reply]
> I’m not sure where the Docker is sitting at the moment, but I updated my workstation yesterday and it’s got Musl 1.2.3 with GCC 11.3.0 for sure:
> 
> ```
> % mipsr5900el-unknown-linux-musl-gcc --version
> mipsr5900el-unknown-linux-musl-gcc (Gentoo 11.3.0 p4) 11.3.0
> Copyright (C) 2021 Free Software Foundation, Inc.
> This is free software; see the source for copying conditions.  There is NO
> warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
> % /usr/mipsr5900el-unknown-linux-musl/usr/lib/libc.so
> musl libc (mipsel-sf)
> Version 1.2.3
> Dynamic Program Loader
> Usage: /usr/mipsr5900el-unknown-linux-musl/usr/lib/libc.so [options] [--] pathname [args]
> ```

Thanks @frno7 this helps a lot!

I'll report this back to the people I'm speaking to but as I have a workaround this can go on the back burner until someone else reports the same problem.

**Reply by frno7** [reply]
My workstation has got `system uname: Linux-5.15.46+-ppc64le-POWER9,_altivec_supported-with-glibc2.34` by the way. Chances are it’s quite unlike yours, unless you have a [POWER9](https://en.wikipedia.org/wiki/POWER9) too? :-)

**Reply by immolo** [reply]
Can I come over and play around your house?

Offtopic for a bit but have you got a model number for that machine so I can look at this further as I have weird hobby of collecting devices with different architectures and this one I don't owe in my collection.

**Reply by frno7** [reply]
It’s a [Talos II](https://www.raptorcs.com/content/TL2B02/intro.html), having 8 cores dual socket, so in total 16 cores (64 hardware threads) with 64 GiB of RAM.

**Reply by immolo** [reply]
I wasn't aware people actually used these outside of a fun project that I've met around the Internet.

If you are ever near me (or visiting my country) then we need go nerd out over a beer or two as I rarely find people as strange as me ;)

**Reply by frno7** [reply]
Sure! Nowadays, when possible, I enjoy free and open computers the most, which include most retro devices of note including the PlayStation 2. For a recent portable device with interesting ideas on modularity, the [DevTerm Kit R-01](https://www.clockworkpi.com/product-page/devterm-kit-r01) looks like a promising starting point, but reviews are [mixed](https://www.talospace.com/2022/05/mini-review-clockwork-pi-devterm-r-01.html). Hopefully there will be improved RISC-V device variants soon.

**Reply by immolo** [reply]
I just love the challenge myself and as this is the first community I've found still active with talented people it's the most fun I've had in the last 2 decades trying to get modern Linux working on an old device.

My labour of love though is trying to get the Free60 community back from the ashes but as none of the lasted kernels build anymore and the last active dev has given up I think the best I can do is bandaid a solution to try and get people back.

I agree with RISC-V though as it's the first CPU that has gotten me excited in decades however it seems to have enough people interested that I think I'd take an itainum up instead as Gentoo are crying out for help before they have to drop support due to lack of interest.

**Comment by immolo** [comment]
Submitted the x86_64 MUSL crossdev possible bug upstream so if we have any other Gentoo users here please add your findings there or here and I'll pass them on to the crossdev and MUSL team in Gentoo.

https://bugs.gentoo.org/852812

**Reply by frno7** [reply]
The relevant error in your build log seems to be `final link failed: Value too large for defined data type`. Could it have something to do with `_FILE_OFFSET_BITS`?

**Reply by immolo** [reply]
I finally moved my main PC over to Gentoo which fixed the issue so it was most likely an Arch issue.

Thanks for the suggestion though
